EIPS/eip-draft_transaction_data_opcodes.md

```diff
@@ -0,0 +1,106 @@
+
+## Motivation
+
+It is undeniable that smart contracts are becoming more interconnected than ever. Up until this point, smart contracts have entirely relied on compliant interfaces and introspection to introduce a new step in the call chain of a complex multi-contract interaction. However, this presents a forwards-only approach which limits the types of interactions that can manifest.
+
+The purpose of this EIP is to provide a way via which a contract is able to identify the entry-point of a transaction on the blockchain and deduce what was the original intention of the transaction by applying introspection on the original transaction data itself.
+
+This EIP enables the development of new types of smart contracts as it can open new pathways for [EIP-721](https://eips.ethereum.org/EIPS/eip-721) NFTs and [EIP-20](https://eips.ethereum.org/EIPS/eip-20) tokens to detect which action their transaction is part of, such as detecting a liquidity provision to a decentralized exchange or a loan within a collateralized lending protocol.
+
+As a side-effect, it acts as an enhanced "substitute" to the `ORIGIN` check that the [EIP-3074](https://eips.ethereum.org/EIPS/eip-3074) breaks, enabling a greater level of granularity to be applied to the introspection of a transaction f.e. by validating the `ENTRYPOINT` being equal to the currently executed contract.
+
+## Specification
+
+### ORIGINDATALOAD (`0x47`), ORIGINDATASIZE (`0x48`) and ORIGINDATACOPY (`0x49`)
+
+These instructions are meant to operate identically to their `CALL`-prefixed counterparts with the exception that they instead operate on the original `data` of a transaction instead of the current call's data. As the data is retrieved once again from the execution environment, the costs for the three instructions will be `G_verylow`, `G_base` and `G_base + G_verylow * (number of words copied, rounded up)` respectively.
+
+### ENTRYPOINT (`0x4a`)
+
+The `ENTRYPOINT` instruction uses 0 stack arguments and pushes the original `to` member of the transaction onto the stack. The address yielded by the instruction is a 160-bit value padded to 256-bits. The operation costs `G_base` to execute, similarly to `ORIGIN` (`0x32`).
+
+### Node Specification
+
+All instructions are meant to provide access to previously inaccessible data that did not exist in the execution context of the EVM. As a result, the data needs to be provided to the execution environments of all contract calls, similarly to how `ORIGIN` data is available as such. A reference implementation will be provided for `go-ethereum` (Golang) that implements the EIP at both the EVM level as well as the node level.
+
+### Compiler Specification
+
+High-level programming languages that compile to EVM bytecode should provide support for these new instructions from a high-level perspective by introducing new members to the transactional context they already support for the `ORIGIN` instruction.
+
```

Comment:
```suggestion
```
Also unnecessary for the specification.  This specification is explicitly for the new EVM opcode(s), and it is up to HLL compiler teams to decide whether they want to add support and how to do so.

Reply:
Again thanks for the feedback, assimilated.